File: data/visualization/cpp/supplier_price.cpp
#include "matplotlibcpp.h"
#include "csv.h"
#include <vector>
#include <string>
#include <map>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <filesystem>
#include <iostream>

namespace plt = matplotlibcpp;

int get_supplier_id(int supplier_id, const std::string& supplier_name) {
    if (supplier_id == 0 && supplier_name.empty()) {
        throw std::runtime_error("Either --id or --name must be specified.");
    }

    if (supplier_id == 0) {
        io::CSVReader<5> in("database/supplier.csv");
        in.read_header(io::ignore_extra_column, "supplier_id", "contact_name", "max_purchase_this_year", "city", "phone_number");
        int id;
        std::string name, city, phone_number;
        int max_purchase_this_year;
        while (in.read_row(id, name, max_purchase_this_year, city, phone_number)) {
            if (name == supplier_name) {
                supplier_id = id;
                break;
            }
        }
        if (supplier_id == 0) {
            throw std::runtime_error("No supplier with name " + supplier_name + " found.");
        }
    }

    return supplier_id;
}

std::map<std::string, double> get_data(int supplier_id) {
    std::map<std::string, double> data;
    io::CSVReader<4> in("database/supply_price_history.csv");
    in.read_header(io::ignore_extra_column, "year", "month", "supplier_id", "price_per_unit");
    int year, month, id;
    double price_per_unit;
    while (in.read_row(year, month, id, price_per_unit)) {
        if (id == supplier_id) {
            std::ostringstream date;
            date << year << "-" << std::setfill('0') << std::setw(2) << month << "-01";
            data[date.str()] = price_per_unit;
        }
    }
    return data;
}

void plot(const std::map<std::string, double>& data, int supplier_id) {
    std::vector<std::string> x;
    std::vector<double> y;
    for (const auto& entry : data) {
        x.push_back(entry.first);
        y.push_back(entry.second);
    }

    plt::plot(x, y, "b-o");
    plt::title("Change in Supply Price for Supplier " + std::to_string(supplier_id));
    plt::xlabel("Date");
    plt::ylabel("Price per Unit");
    plt::grid(true);
    plt::tight_layout();

    std::filesystem::create_directories("images");
    plt::save("images/supplier_price_" + std::to_string(supplier_id) + ".png");
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " --id <supplier_id> or " << argv[0] << " --name <supplier_name>" << std::endl;
        return 1;
    }

    int supplier_id = 0;
    std::string supplier_name;
    if (std::string(argv[1]) == "--id") {
        supplier_id = std::stoi(argv[2]);
    } else if (std::string(argv[1]) == "--name") {
        supplier_name = argv[2];
    } else {
        std::cerr << "Invalid option: " << argv[1] << std::endl;
        return 1;
    }

    supplier_id = get_supplier_id(supplier_id, supplier_name);
    auto data = get_data(supplier_id);
    plot(data, supplier_id);

    return 0;
}

File: opti_coffee/visualization/supplier_price.py
import pandas as pd
import matplotlib.pyplot as plt
import argparse
import os

def get_args():
    parser = argparse.ArgumentParser(description='Plot the change in supply price for a given supplier.')
    
    parser.add_argument('--id', type=int, default=None, help='Supplier ID')
    parser.add_argument('--name', type=str, default=None, help='Supplier Name')
    
    return parser.parse_args()

def get_supplier_id(args):
    if args.id is None and args.name is None:
        raise ValueError('Either --id or --name must be specified.')

    if args.id is None:
        # Get the supplier ID from the name
        suppliers = pd.read_csv('database/supplier.csv')
        rows = (suppliers['contact_name'] == args.name)
        
        if sum(rows) == 0:
            raise ValueError(f'No supplier with name {args.name} found.')
        elif sum(rows) > 1:
            raise ValueError(f'Multiple suppliers with name {args.name} found.')
        idx = rows.idxmax()
        supplier_id = suppliers.loc[idx, 'supplier_id']
    else:
        supplier_id = args.id
    
    return supplier_id

def get_data(supplier_id):
    # Load the data
    df = pd.read_csv('database/supply_price_history.csv')

    # Filter data for the given supplier
    df_filtered = df[df['supplier_id'] == supplier_id]

    # Combine 'year' and 'month' columns and convert to datetime
    df_filtered['date'] = pd.to_datetime(df_filtered[['year', 'month']].assign(DAY=1))

    # Sort by date
    return df_filtered.sort_values(by='date')

def plot(filtered_data, supplier_id):
    # Plotting
    plt.plot(filtered_data['date'], filtered_data['price_per_unit'], marker='o', color='b', linestyle='-')
    plt.title(f'Change in Supply Price for Supplier {supplier_id}')
    plt.xlabel('Date')
    plt.ylabel('Price per Unit')
    plt.grid(True)
    plt.tight_layout()

    # Show the plot
    os.makedirs('images', exist_ok=True)
    plt.savefig(f'images/supplier_price_{supplier_id}.png')

if __name__ == '__main__':
    args = get_args()
    supplier_id = get_supplier_id(args)
    filtered_data = get_data(supplier_id)
    plot(filtered_data, supplier_id)
